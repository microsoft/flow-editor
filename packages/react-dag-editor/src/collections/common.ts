export type Owner = number;

export interface IMapBuilder<K, V> {
  get(key: K): V | undefined;
  has(key: K): boolean;
  set(key: K, value: V): IMapBuilder<K, V>;
  update(key: K, updater: (prev: V) => V): IMapBuilder<K, V>;
  delete(key: K): IMapBuilder<K, V>;
  finish(): IMap<K, V>;
}

export interface IMap<K, V> extends Iterable<[K, V]> {
  size: number;
  get(key: K): V | undefined;
  has(key: K): boolean;
  set(key: K, value: V): IMap<K, V>;
  update(key: K, updater: (prev: V) => V): IMap<K, V>;
  delete(key: K): IMap<K, V>;
  clone(): IMap<K, V>;
  entries(): IterableIterator<[K, V]>;
  values(): IterableIterator<V>;
  mutate(): IMapBuilder<K, V>;
  map<T>(f: (value: V) => T): IMap<K, T>;
  filter<T>(predicate: (value: V, key: K) => boolean): IMap<K, V>;
  forEach(f: (value: V, key: K) => void): void;
  find(predicate: (value: V) => boolean): V | undefined;
}
