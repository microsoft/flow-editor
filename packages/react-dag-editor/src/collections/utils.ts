export const is = Object.is;

export class MapIterator<T, U> implements IterableIterator<U> {
  private readonly upstream: Iterator<T>;
  private readonly f: (value: T) => U;

  public constructor(upstream: Iterator<T>, f: (value: T) => U) {
    this.upstream = upstream;
    this.f = f;
  }

  public [Symbol.iterator](): IterableIterator<U> {
    return this;
  }

  public next(): IteratorResult<U> {
    const next = this.upstream.next();
    if (next.done) {
      return next;
    }
    return {
      done: false,
      value: this.f(next.value)
    };
  }
}
