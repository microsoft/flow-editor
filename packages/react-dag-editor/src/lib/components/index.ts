export * from "./ReactDagEditor";
export * from "./Graph/Graph";
export * from "./ItemPanel";
export * from "./Minimap/Minimap";
export * from "./StaticGraph/StaticGraph";
export * from "./Graph/IGraphProps";
export * from "./RegisterComponent";
export { defaultGetPositionFromEvent } from "../controllers";
export * from "./GridBackground";
export * from "./NodeAnchors";
