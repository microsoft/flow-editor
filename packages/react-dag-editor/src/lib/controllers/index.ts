export * from "./DragController";
export * from "./DragNodeController";
export * from "./TouchController";
export * from "./TwoFingerHandler";
