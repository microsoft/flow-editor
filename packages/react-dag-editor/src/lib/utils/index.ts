export * from "./getPortPosition";
export * from "./getNodeConfig";
export * from "./graphDataUtils";
export * from "./transformMatrix";
export * from "./getCurvePathD";
export * from "./layout";
export * from "./browsers";
export * from "./isWithinThreshold";
export * from "./geometric";
export * from "./viewport";
export * from "./Writable";
export * from "./gantt/getCriticalPath";
export * from "./gantt/getTopoSortingNodes";
export * from "./getOffsetLimit";
export * from "./event";
export * from "./valueOrCallback";
export * from "./debug";
export * from "./debounce";
export * from "./arrayEquals";
export * from "./empty";
export * from "./stopPropagation";
export * from "./clamp";
export * from "./history";
export * as Bitset from "./bitset";
