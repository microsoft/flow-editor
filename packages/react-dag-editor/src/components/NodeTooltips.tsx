import * as React from "react";
import { useGraphConfig } from "../hooks/context";
import type { IViewport } from "../models/geometry";
import type { NodeModel } from "../models/NodeModel";
import { GraphNodeStatus } from "../models/status";
import { getNodeConfig } from "../utils";
import * as Bitset from "../utils/bitset";

interface IProps {
  node?: NodeModel;
  viewport: Required<IViewport>;
}

export const NodeTooltips: React.FunctionComponent<IProps> = (props) => {
  const { node, viewport } = props;
  const graphConfig = useGraphConfig();

  if (!node) {
    return null;
  }

  if (!Bitset.has(GraphNodeStatus.Activated)(node.status)) {
    return null;
  }

  const nodeConfig = getNodeConfig(node, graphConfig);

  if (!nodeConfig?.renderTooltips) {
    return null;
  }

  return (
    <div className="node-tooltips">
      {nodeConfig.renderTooltips({
        model: node,
        viewport,
      })}
    </div>
  );
};
