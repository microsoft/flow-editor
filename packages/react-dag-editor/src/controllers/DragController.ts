/* eslint-disable no-invalid-this */
import { IEventProvider, IGlobalMoveEventTypes } from "../event-provider/types";
import { IPoint } from "../models/geometry";
import { noop } from "../utils/noop";
import { animationFramed, IScheduledCallback } from "../utils/scheduling";

export type TGetPositionFromEvent = (e: MouseEvent) => IPoint;

export interface IOnDragMove<T = MouseEvent> {
  /**
   * clientX after getPositionFromEvent
   */
  clientX: number;
  /**
   * clientY after getPositionFromEvent
   */
  clientY: number;
  dx: number;
  dy: number;
  totalDX: number;
  totalDY: number;
  e: T;
}

export interface IOnDragEnd<T = MouseEvent> {
  totalDX: number;
  totalDY: number;
  e: T;
}

export class DragController<T extends MouseEvent = MouseEvent> {
  public onMove: (move: IOnDragMove) => void = noop;
  public onEnd: (end: IOnDragEnd) => void = noop;
  protected mouseMove: IScheduledCallback<[T]>;
  protected lastEvent: T | null = null;
  private readonly getPositionFromEvent: TGetPositionFromEvent;
  private startX = 0;
  private startY = 0;
  private prevClientX = 0;
  private prevClientY = 0;
  private readonly eventProvider: IEventProvider<IGlobalMoveEventTypes>;

  public constructor(
    eventProvider: IEventProvider<IGlobalMoveEventTypes>,
    getPositionFromEvent: TGetPositionFromEvent
  ) {
    this.eventProvider = eventProvider;
    this.getPositionFromEvent = getPositionFromEvent;
    this.mouseMove = animationFramed((e: T) => {
      this.doOnMouseMove(e);
    });
  }

  public start(e: T): void {
    this.lastEvent = e;
    const { x, y } = this.getPositionFromEvent(e);
    this.startX = x;
    this.startY = y;
    this.prevClientX = x;
    this.prevClientY = y;

    this.eventProvider.on("move", this.onMouseMove as (e: MouseEvent) => void);
    this.eventProvider.on("end", this.onMouseUp as (e: MouseEvent) => void);
  }

  public stop(): void {
    this.mouseMove.cancel();
    this.eventProvider.off("move", this.onMouseMove as (e: MouseEvent) => void);
    this.eventProvider.off("end", this.onMouseUp as (e: MouseEvent) => void);
  }

  protected getDelta(x: number, y: number): IPoint {
    const clientDX = x - this.prevClientX;
    const clientDY = y - this.prevClientY;
    this.prevClientX = x;
    this.prevClientY = y;
    return {
      x: clientDX,
      y: clientDY,
    };
  }

  protected getTotalDelta(e: MouseEvent): IPoint {
    const x = e.clientX - this.startX;
    const y = e.clientY - this.startY;
    return {
      x,
      y,
    };
  }

  protected doOnMouseMove(e: MouseEvent): void {
    const { x, y } = this.getPositionFromEvent(e);
    const { x: dx, y: dy } = this.getDelta(x, y);
    const { x: totalDX, y: totalDY } = this.getTotalDelta(e);
    this.onMove({
      clientX: x,
      clientY: y,
      dx,
      dy,
      totalDX,
      totalDY,
      e,
    });
  }

  protected doOnMouseUp(e: T): void {
    e.preventDefault();
    const { x: totalDX, y: totalDY } = this.getTotalDelta(e);
    this.onEnd({ totalDX, totalDY, e });

    this.stop();
  }

  private readonly onMouseUp = (e: T) => {
    this.lastEvent = e;
    this.doOnMouseUp(e);
    this.lastEvent = null;
  };

  private readonly onMouseMove = (e: T) => {
    this.lastEvent = e;
    e.preventDefault();
    this.mouseMove(e);
  };
}

export function defaultGetPositionFromEvent<T extends MouseEvent>(
  e: T
): IPoint {
  return {
    x: e.clientX,
    y: e.clientY,
  };
}
