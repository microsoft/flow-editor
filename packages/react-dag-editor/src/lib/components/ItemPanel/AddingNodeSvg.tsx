import * as React from "react";
import { v4 as uuid } from "uuid";
import { useConst } from "../../hooks/useConst";
import {
  useAlignmentLines,
  useGraphConfig,
  useGraphController,
  useViewport,
} from "../../hooks/context";
import { GraphCanvasEvent } from "../../models/event";
import { ICanvasNode } from "../../models/node";
import { NodeModel } from "../../models/NodeModel";
import { getNodeSize } from "../../utils";
import { defaultGetNodeAriaLabel } from "../../utils/a11yUtils";
import { getAutoAlignDisplacement } from "../../utils/autoAlign";
import { EventChannel } from "../../utils/eventChannel";
import { GraphNode } from "../GraphNode";
import classes from "../Graph.styles.module.scss";
import { Transform } from "../Transform";
import { useSvgRect } from "./useSvgRect";

interface IAddingNodeSvgProps<NodeData, PortData> {
  model: ICanvasNode<NodeData, PortData>;
  nextNodeRef: React.MutableRefObject<ICanvasNode<NodeData, PortData> | null>;
  svgRef: React.RefObject<SVGSVGElement>;
}

export const AddingNodeSvg: React.FunctionComponent<
  IAddingNodeSvgProps<unknown, unknown>
> = ({ model, svgRef, nextNodeRef }) => {
  const rect = useSvgRect(svgRef);
  const graphConfig = useGraphConfig();
  const graphController = useGraphController();
  const alignmentLines = useAlignmentLines();
  const viewport = useViewport();

  const dummyNode = React.useMemo(() => {
    if (!model || !viewport.rect) {
      return null;
    }
    const { transformMatrix } = viewport;
    const { left, top } = viewport.rect;
    const diffLeft = left / transformMatrix[0];
    const diffTop = top / transformMatrix[3];
    return {
      id: model.id,
      x: model.x - diffLeft,
      y: model.y - diffTop,
      ...getNodeSize(model, graphConfig),
    };
  }, [graphConfig, viewport, model]);

  React.useLayoutEffect((): void => {
    if (!dummyNode) {
      return;
    }
    graphController.eventChannel.trigger({
      type: GraphCanvasEvent.DraggingNodeFromItemPanel,
      node: dummyNode,
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [dummyNode?.x, dummyNode?.y]);

  const attachedNode = React.useMemo<ICanvasNode | null>(() => {
    if (!dummyNode || !model) {
      return null;
    }

    const dxAligned = getAutoAlignDisplacement(
      alignmentLines,
      [dummyNode],
      graphConfig,
      "x"
    );
    const dyAligned = getAutoAlignDisplacement(
      alignmentLines,
      [dummyNode],
      graphConfig,
      "y"
    );

    return {
      ...model,
      x: model.x + dxAligned,
      y: model.y + dyAligned,
    };
  }, [alignmentLines, dummyNode, graphConfig, model]);

  React.useEffect(() => {
    nextNodeRef.current = attachedNode;
  }, [nextNodeRef, attachedNode]);

  const eventChannel = useConst(() => new EventChannel());

  const tempGraphId = useConst(uuid);

  const node = React.useMemo(
    () => NodeModel.fromJSON(attachedNode ?? model, undefined, undefined),
    [attachedNode, model]
  );

  return (
    <svg
      id={tempGraphId}
      ref={svgRef}
      className={classes["adding-node-svg"]}
      preserveAspectRatio="xMidYMid meet"
    >
      <Transform matrix={viewport.transformMatrix}>
        {rect && (
          <GraphNode
            graphId={tempGraphId}
            node={node}
            viewport={{
              rect,
              transformMatrix: viewport.transformMatrix,
            }}
            eventChannel={eventChannel}
            getNodeAriaLabel={defaultGetNodeAriaLabel}
          />
        )}
      </Transform>
    </svg>
  );
};
