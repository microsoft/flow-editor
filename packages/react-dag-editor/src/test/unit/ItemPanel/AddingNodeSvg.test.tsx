/* eslint-disable @typescript-eslint/no-explicit-any */
import {
  act,
  cleanup,
  fireEvent,
  render,
  RenderResult,
  screen,
} from "@testing-library/react";
import * as React from "react";
import * as ShallowRenderer from "react-test-renderer/shallow";
import { GraphCanvasEvent, GraphConfigBuilder, rect } from "../../../index";
import { MouseEventButton } from "../../../lib/common/constants";
import { Item } from "../../../lib/components/ItemPanel";
import { AddingNodeSvg } from "../../../lib/components/ItemPanel/AddingNodeSvg";
import { GraphController } from "../../../lib/controllers/GraphController";
import { GraphControllerRef, TestComponent } from "../../TestComponent";
import { mockClientRect, patchPointerEvent } from "../../utils";
import { TestItemContent } from "./TestItemContent";

jest.mock("../../../lib/components/ItemPanel/useSvgRect", () => ({
  useSvgRect: () => {
    return mockClientRect;
  },
}));

describe("ItemPanel - AddingNodeSvg", () => {
  let renderedWrapper: RenderResult;
  let graphController: GraphController;
  const getElement = () => renderedWrapper.getByRole("button");

  beforeAll(() => {
    patchPointerEvent();
    jest.useFakeTimers();
  });
  afterAll(() => {
    jest.clearAllMocks();
  });

  beforeEach(() => {
    const graphConfig = GraphConfigBuilder.default()
      .registerNode("nodeShape", rect)
      .build();
    const graphControllerRef = React.createRef<GraphController>();
    const getNode = () => {
      return {
        name: "node1",
        shape: "nodeShape",
      };
    };
    renderedWrapper = render(
      <TestComponent graph={false} settings={{ graphConfig }}>
        <Item getNode={getNode} dragWillStart={jest.fn()}>
          <TestItemContent text="test item for addingNodeSVG" />
        </Item>
        <GraphControllerRef ref={graphControllerRef} />
      </TestComponent>
    );
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    graphController = graphControllerRef.current!;
    expect(graphController).toBeDefined();
    act(() => {
      graphController.dispatch({
        type: GraphCanvasEvent.ViewportResize,
        viewportRect: mockClientRect,
      });
    });
  });

  afterEach(cleanup);

  /**
   * @param el
   * @param mouseEventButton
   * @param points
   */
  function simulateDragging(
    el: HTMLElement | Window,
    mouseEventButton: MouseEventButton,
    points: Array<[number, number]>
  ): void {
    if (points.length === 0) {
      return;
    }

    const first = points[0];
    act(() => {
      fireEvent.pointerDown(el, {
        clientX: first[0],
        clientY: first[1],
        button: mouseEventButton,
        pointerType: "mouse",
      });
      jest.runAllTimers();
    });

    points.forEach(([clientX, clientY]) => {
      act(() => {
        fireEvent.pointerMove(window, {
          clientX,
          clientY,
          button: mouseEventButton,
          pointerType: "mouse",
        });
      });
    });

    const last = points[points.length - 1];
    act(() => {
      jest.runAllTimers();
      fireEvent.pointerUp(window, {
        clientX: last[0],
        clientY: last[1],
        button: mouseEventButton,
        pointerType: "mouse",
      });
    });
  }

  it("Should match the snapshot", () => {
    const renderer = ShallowRenderer.createRenderer();
    renderer.render(
      <AddingNodeSvg
        nextNodeRef={{ current: null }}
        model={{
          name: "node1",
          shape: "nodeShape",
          x: 0,
          y: 0,
          id: "mock-id",
        }}
        svgRef={React.createRef<SVGSVGElement>()}
      />
    );
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("adding node", () => {
    simulateDragging(screen.getByRole("button"), MouseEventButton.Primary, [
      [100, 100],
      [105, 105],
    ]);

    expect(getElement()).toMatchSnapshot();
  });

  it("not adding node if out of graph", () => {
    jest.spyOn(Element.prototype, "getBoundingClientRect").mockReturnValue({
      top: 50,
      right: 200,
      bottom: 200,
      left: 50,
      width: 150,
      height: 150,
      x: 50,
      y: 50,
    } as any);

    simulateDragging(screen.getByRole("button"), MouseEventButton.Primary, [
      [25, 25],
      [30, 30],
    ]);

    expect(getElement()).toMatchSnapshot();
  });

  it("without left mouse down when dragging", () => {
    simulateDragging(screen.getByRole("button"), MouseEventButton.Secondary, [
      [100, 100],
      [105, 105],
    ]);

    expect(getElement()).toMatchSnapshot();
  });
});
