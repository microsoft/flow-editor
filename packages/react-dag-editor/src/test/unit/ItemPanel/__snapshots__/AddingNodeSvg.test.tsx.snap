// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemPanel - AddingNodeSvg Should match the snapshot 1`] = `
<svg
  className="addingNodeSvg"
  id="mock-uuid"
  preserveAspectRatio="xMidYMid meet"
>
  <Transform
    matrix={
      Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ]
    }
  >
    <GraphNode
      eventChannel={
        EventChannel {
          "externalHandlerRef": Object {
            "current": null,
          },
          "listenersRef": Object {
            "current": null,
          },
          "queue": Array [],
          "working": false,
        }
      }
      getNodeAriaLabel={[Function]}
      graphId="mock-uuid"
      node={
        Object {
          "ariaLabel": undefined,
          "automationId": undefined,
          "height": undefined,
          "id": "mock-id",
          "name": "node1",
          "ports": undefined,
          "properties": Object {},
          "shape": "nodeShape",
          "status": undefined,
          "width": undefined,
          "x": 0,
          "y": 0,
        }
      }
      viewport={
        Object {
          "rect": [Circular],
          "transformMatrix": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
        }
      }
    />
  </Transform>
</svg>
`;

exports[`ItemPanel - AddingNodeSvg Should match the snapshot: events 1`] = `Array []`;

exports[`ItemPanel - AddingNodeSvg adding node 1`] = `
<div
  class="moduleItem"
  role="button"
>
  <p>
    test item for addingNodeSVG:default
  </p>
</div>
`;

exports[`ItemPanel - AddingNodeSvg adding node: events 1`] = `Array []`;

exports[`ItemPanel - AddingNodeSvg not adding node if out of graph 1`] = `
<div
  class="moduleItem"
  role="button"
>
  <p>
    test item for addingNodeSVG:default
  </p>
</div>
`;

exports[`ItemPanel - AddingNodeSvg not adding node if out of graph: events 1`] = `Array []`;

exports[`ItemPanel - AddingNodeSvg without left mouse down when dragging 1`] = `
<div
  class="moduleItem"
  role="button"
>
  <p>
    test item for addingNodeSVG:default
  </p>
</div>
`;

exports[`ItemPanel - AddingNodeSvg without left mouse down when dragging: events 1`] = `Array []`;
